### PR TITLE
Implement Effect.fn to define traced functions

### PR DESCRIPTION
**Pull request by mikearnaldi**
Allows:

```ts
import { Effect } from "effect"

const logExample = Effect.fn("logExample")(
  function*<N extends number>(n: N) {
    yield* Effect.annotateCurrentSpan("n", n)
    yield* Effect.logInfo(`got: ${n}`)
    yield* Effect.fail(new Error())
  },
  Effect.delay("1 second")
)

Effect.runFork(
  logExample(100).pipe(
    Effect.catchAllCause(Effect.logError)
  )
)
```

### CODE REVIEW

File: packages/effect/src/Effect.ts [hunk 1]
```diff
@@ -6912,3 +6912,323 @@ export declare namespace Service {
...
+        ? gen(() => internalCall(() => body.apply(self, args)))
+        : body.apply(self, args)
+      for (const x of pipeables) {
+        res = x(res)
+      }
+      return res
+    }
+    const opts: any = (options && "captureStackTrace" in options) ? options : { captureStackTrace, ...options }
+    return withSpan(
+      core.suspend(() => res(...args)),
```
Comment (tim-smart) [inline packages/effect/src/Effect.ts:7229]:
```suggestion
    const opts: any = (options && "captureStackTrace" in options) ? options : { captureStackTrace, ...options }
    let effect: Effect<any, any, any>
    return withSpan(
      core.suspend(() => effect ??= res(...args)),
```

Could cache the evaluation here.

**Review by tim-smart — COMMENTED** [review]
Seems good to go. Left one comment.

File: packages/effect/src/Effect.ts (hunk 1 above)
Comment (mikearnaldi) [reply packages/effect/src/Effect.ts:7229]: `res` isn't assumed to be pure so caching would lead to potentially different results

**Comment by mikearnaldi** [comment]
in theory if we make this compatible conditionally with every function it could be auto-applied via a babel-plugin, will think about it a bit more

File: packages/effect/src/Effect.ts (hunk 1 above)
Comment (tim-smart) [reply packages/effect/src/Effect.ts:7229]: I was think to perserve the semantics of calling a function:

```ts
const someFn = Effect.fn("someFn")(...)

// currently
const effect = someFn(...)

// function will be called once per iteration
// when really it should only be called once
Effect.forever(effect)
```

Comment (mikearnaldi) [reply packages/effect/src/Effect.ts:7229]: not sure we can, if the function is specified using a generator the whole body may contain side-effects and repeating the effect is expected to repeat the side-effects

Comment (tim-smart) [reply packages/effect/src/Effect.ts:7229]: Ah yeah, we can only do it for the non-generator case. Just noticed a small bug where pipeables like `Effect.catchAllCause` would not catch defects from calling the function. This should fix it:

```diff
diff --git a/packages/effect/src/Effect.ts b/packages/effect/src/Effect.ts
index a2b4c2cc3..48f1057d6 100644
--- a/packages/effect/src/Effect.ts
+++ b/packages/effect/src/Effect.ts
@@ -7215,20 +7215,18 @@ export const fn: (
     // @ts-ignore
     // eslint-disable-next-line @typescript-eslint/no-this-alias
     const self = this
-    const res = (...args: Array<any>) => {
-      let res = isGeneratorFunction(body)
+    let effect: Effect<any, any, any>
+    try {
+      effect = isGeneratorFunction(body)
         ? gen(() => internalCall(() => body.apply(self, args)))
         : body.apply(self, args)
-      for (const x of pipeables) {
-        res = x(res)
-      }
-      return res
+    } catch (defect) {
+      effect = die(defect)
+    }
+    for (const x of pipeables) {
+      effect = x(effect)
     }
     const opts: any = (options && "captureStackTrace" in options) ? options : { captureStackTrace, ...options }
-    return withSpan(
-      core.suspend(() => res(...args)),
-      name,
-      opts
-    )
+    return withSpan(effect, name, opts)
   }
 }
```

Comment (mikearnaldi) [reply packages/effect/src/Effect.ts:7229]: hmm this has the same issue if a defect arises in a pipeable, wondering if we shouldn't just suspend everything

Comment (tim-smart) [reply packages/effect/src/Effect.ts]: I think I have the behaviour correct now.

**Review by tim-smart — APPROVED** [review]

**Pull request merged**